File: tobas_setup_assistant/src/tobas_setup_assistant/setting_widgets/gps.py
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ..setup_assistant import SetupAssistant

from overrides import override

from ..common import SENSOR_OFFSET_DESCRIPTION
from ..parameter_getters import ParamGetterWidget_SpinBox, ParamGetterWidget_DoubleSpinBox, ParamGetterWidget_Vector3d
from .base_setting import OptionalDeviceWidget


class GpsWidget(OptionalDeviceWidget):
    NAME = "GPS"

    def __init__(self, main: SetupAssistant) -> None:
        title_text = "Define Global Positioning System"
        abst_text = ""  # TODO
        super().__init__(main, title_text, abst_text, True)

        self.offset = ParamGetterWidget_Vector3d("Offset", SENSOR_OFFSET_DESCRIPTION, suffix=" m")
        self._add_config_widget(self.offset)

        update_rate_description = ""  # TODO
        self.update_rate = ParamGetterWidget_SpinBox(
            "Update rate", update_rate_description, minimum=1, default=5, suffix=" Hz"
        )
        self._add_config_widget(self.update_rate)

        delay_description = ""  # TODO
        self.delay = ParamGetterWidget_DoubleSpinBox(
            "Communication delay", delay_description, decimals=2, minimum=0.0, default=0.2, suffix=" s"
        )
        self._add_config_widget(self.delay)

        pos_corr_time_description = ""  # TODO
        self.pos_corr_time = ParamGetterWidget_SpinBox(
            "Position correction time constant", pos_corr_time_description, minimum=1, default=10, suffix=" s"
        )
        self._add_config_widget(self.pos_corr_time)

        horizontal_pos_accuracy_description = ""  # TODO
        self.horizontal_pos_accuracy = ParamGetterWidget_DoubleSpinBox(
            "Horizontal position accuracy",
            horizontal_pos_accuracy_description,
            decimals=2,
            minimum=0.0,
            default=2.0,
            suffix=" m",
        )
        self._add_config_widget(self.horizontal_pos_accuracy)

        vertical_pos_accuracy_description = ""  # TODO
        self.vertical_pos_accuracy = ParamGetterWidget_DoubleSpinBox(
            "Hertical position accuracy",
            vertical_pos_accuracy_description,
            decimals=2,
            minimum=0.0,
            default=4.0,
            suffix=" m",
        )
        self._add_config_widget(self.vertical_pos_accuracy)

        horizontal_vel_stddev_description = ""  # TODO
        self.horizontal_vel_stddev = ParamGetterWidget_DoubleSpinBox(
            "Standard deviation for horizontal speed noise",
            horizontal_vel_stddev_description,
            decimals=2,
            minimum=0.0,
            default=0.1,
            suffix=" m/s",
        )
        self._add_config_widget(self.horizontal_vel_stddev)

        vertical_vel_stddev_description = ""  # TODO
        self.vertical_vel_stddev = ParamGetterWidget_DoubleSpinBox(
            "Standard deviation for vertical speed noise",
            vertical_vel_stddev_description,
            decimals=2,
            minimum=0.0,
            default=0.1,
            suffix=" m/s",
        )
        self._add_config_widget(self.vertical_vel_stddev)

        self._rows.addStretch()

    @override
    def update_internal_data_structures(self) -> None:
        pass

    @override
    def is_valid(self) -> bool:
        if not self.equipped():
            return True

        return True

File: tobas_gui/tobas_setup_assistant/src/setting_widgets/gps.cpp
#include <tobas_yaml_tools/convert/eigen.hpp>

#include "tobas_setup_assistant/setting_tabs/gps.hpp"

namespace gui
{
namespace setup_assistant
{
GPSWidget::GPSWidget()
{
  offset_ = new ParamGetterWidget_Vector3d("Offset", kSensorOffsetDescription);
  offset_->setSuffix(" m");
  addParamWidget(offset_);

  update_rate_ = new ParamGetterWidget_SpinBox("Update Rate", "");  // TODO
  update_rate_->setMinimum(1);
  update_rate_->setValue(5);
  update_rate_->setSuffix(" Hz");
  addParamWidget(update_rate_);

  delay_ = new ParamGetterWidget_DoubleSpinBox("Communication Delay", "");  // TODO
  delay_->setDecimals(2);
  delay_->setMinimum(0.);
  delay_->setValue(0.2);
  delay_->setSuffix(" s");
  addParamWidget(delay_);

  pos_corr_time_ = new ParamGetterWidget_SpinBox("Position Correction Time Constant", "");  // TODO
  pos_corr_time_->setMinimum(1);
  pos_corr_time_->setValue(10);
  pos_corr_time_->setSuffix(" s");
  addParamWidget(pos_corr_time_);

  horizontal_pos_accuracy_ = new ParamGetterWidget_DoubleSpinBox("Horizontal Position Accuracy", "");  // TODO
  horizontal_pos_accuracy_->setDecimals(2);
  horizontal_pos_accuracy_->setMinimum(0.);
  horizontal_pos_accuracy_->setValue(2.);
  horizontal_pos_accuracy_->setSuffix(" m");
  addParamWidget(horizontal_pos_accuracy_);

  vertical_pos_accuracy_ = new ParamGetterWidget_DoubleSpinBox("Vertical Position Accuracy", "");  // TODO
  vertical_pos_accuracy_->setDecimals(2);
  vertical_pos_accuracy_->setMinimum(0.);
  vertical_pos_accuracy_->setValue(4.);
  vertical_pos_accuracy_->setSuffix(" m");
  addParamWidget(vertical_pos_accuracy_);

  horizontal_vel_stddev_ =
    new ParamGetterWidget_DoubleSpinBox("Standard Deviation for Horizontal Speed Noise", "");  // TODO
  horizontal_vel_stddev_->setDecimals(2);
  horizontal_vel_stddev_->setMinimum(0.);
  horizontal_vel_stddev_->setValue(0.1);
  horizontal_vel_stddev_->setSuffix(" m/s");
  addParamWidget(horizontal_vel_stddev_);

  vertical_vel_stddev_ =
    new ParamGetterWidget_DoubleSpinBox("Standard Deviation for Vertical Speed Noise", "");  // TODO
  vertical_vel_stddev_->setDecimals(2);
  vertical_vel_stddev_->setMinimum(0.);
  vertical_vel_stddev_->setValue(0.1);
  vertical_vel_stddev_->setSuffix(" m/s");
  addParamWidget(vertical_vel_stddev_);

  addStretch();
}

const char* GPSWidget::name() const
{
  return "GPS";
}

const char* GPSWidget::title() const
{
  return "Define Global Positioning System";
}

const char* GPSWidget::description() const
{
  return "";  // TODO
}

void GPSWidget::onOpened()
{
  return;
}

void GPSWidget::updateInternalDataStructures()
{
  return;
}

bool GPSWidget::isValid()
{
  if (!equipped())
    return true;

  return true;
}

YAML::Node GPSWidget::dump()
{
  YAML::Node node(YAML::NodeType::Map);

  node[kEquippedKey] = equipped_->isChecked();

  node[offset_->name()] = offset_->getValue();
  node[update_rate_->name()] = update_rate_->getValue();
  node[delay_->name()] = delay_->getValue();
  node[pos_corr_time_->name()] = pos_corr_time_->getValue();
  node[horizontal_pos_accuracy_->name()] = horizontal_pos_accuracy_->getValue();
  node[vertical_pos_accuracy_->name()] = vertical_pos_accuracy_->getValue();
  node[horizontal_vel_stddev_->name()] = horizontal_vel_stddev_->getValue();
  node[vertical_vel_stddev_->name()] = vertical_vel_stddev_->getValue();

  return node;
}

void GPSWidget::load(const YAML::Node& node)
{
  equipped_->setChecked(node[kEquippedKey].as<bool>());

  offset_->setValue(node[offset_->name()].as<Eigen::Vector3d>());
  update_rate_->setValue(node[update_rate_->name()].as<int>());
  delay_->setValue(node[delay_->name()].as<double>());
  pos_corr_time_->setValue(node[pos_corr_time_->name()].as<int>());
  horizontal_pos_accuracy_->setValue(node[horizontal_pos_accuracy_->name()].as<double>());
  vertical_pos_accuracy_->setValue(node[vertical_pos_accuracy_->name()].as<double>());
  horizontal_vel_stddev_->setValue(node[horizontal_vel_stddev_->name()].as<double>());
  vertical_vel_stddev_->setValue(node[vertical_vel_stddev_->name()].as<double>());
}

Eigen::Vector3d GPSWidget::offset() const
{
  return offset_->getValue();
}

int GPSWidget::updateRate() const
{
  return update_rate_->getValue();
}

double GPSWidget::delay() const
{
  return delay_->getValue();
}

int GPSWidget::positionCorrectionTime() const
{
  return pos_corr_time_->getValue();
}

double GPSWidget::horizontalPositionAccuracy() const
{
  return horizontal_pos_accuracy_->getValue();
}

double GPSWidget::verticalPositionAccuracy() const
{
  return vertical_pos_accuracy_->getValue();
}

double GPSWidget::horizontalVelocityStddev() const
{
  return horizontal_vel_stddev_->getValue();
}

double GPSWidget::verticalVelocityStddev() const
{
  return vertical_vel_stddev_->getValue();
}

bool GPSWidget::defaultEquipped() const
{
  return true;
}
}  // namespace setup_assistant
}  // namespace gui
